### ISSUE TITLE
userspace: naive header parsing algorithm in parse_syscalls.py

### ISSUE THREAD

**Issue by andrewboie** [zephyrproject-rtos/zephyr#14592]
**Is your enhancement proposal related to a problem? Please describe.**
parse_syscalls.py has a naive algorithm:

- It very naively parses system call invocations. It can't handle argument type declarations which are arrays, or function pointers.
- It looks at headers without invoking the preprocessor. This causes every `__syscall` found to generate an 8-byte entry in the system call dispatch table, even if the syscall is unimplemented. This situation is handled gracefully at runtime if such a call is invoked, but space is wasted in the table.
- Another side effect of not invoking the preprocessor is that you can't conditionally define whether an API is a syscall, for example:

```
  #ifdef FOO_NEEDS_THESE_SYSCALLS
  #define FOO_SYSCALL __syscall
  #else
  #define FOO_SYSCALL
  #endif

  FOO_SYSCALL foo_yourself(void);
```
Hopelessly confuses it.

**Describe the solution you'd like**
1. The semantics of __syscall should work, from an end user perspective, like any other C token
2. The full extent of the C language should be usable for syscall function protoypes.
3. Unimplemented system calls don't take up room in the dispatch table\

Number 3 is what really annoys me. This is overall a low-priority issue though.

I thought of something clever of using DWARF parsing to get the syscall metadata instead of parsing headers, but this has a chicken-egg problem in that the generated syscall headers need to be in place or nothing compiles.

**Comment by zephyrbot** [comment]
Hi @dcpleung,

This issue, marked as an Enhancement, was opened a while ago and did not get any traction. It was just assigned to you based on the labels. If you don't consider yourself the right person to address this issue, please re-assing it to the right person.

Please take a moment to review if the issue is still relevant to the project.  If it is, please provide feedback and direction on how to move forward. If it is not, has already been addressed, is a duplicate, or is no longer relevant, please close it with a short comment explaining the reason.

@andrewboie you are also encouraged to help moving this issue forward by providing additional information and confirming this request/issue is still relevant to you.

Thanks!

**Comment by dcpleung** [comment]
Number 3 is partially fixed by https://github.com/zephyrproject-rtos/zephyr/pull/58351 as not all syscalls are included.

The chicken-egg problem of needing syscalls headers to do a first pass before we can re-parse syscalls for actual arguments can't really be solved without adding lots of complexities in the build system. This would at least require adding another compilation pass to the mix. So I guess for now, we keep the current mechanism until there is a really strong demand for changing.

Closing now...
